File: backend/app/routes/simulations.py
from fastapi import APIRouter, HTTPException
from app.repositories.flow_repository import FlowRepository
from app.repositories.simulation_repository import SimulationRepository
from app.repositories.trace_repository import TraceRepository
from app.schemas.comparison import CompareRequest, CompareResponse
from app.schemas.simulation import RunSimulationRequest, Scenario, SimulationRun
from app.services.simulation_service import SimulationService

router = APIRouter(tags=["simulations"])


def _get_service() -> SimulationService:
    return SimulationService(FlowRepository(), SimulationRepository(), TraceRepository())


@router.post(
    "/api/projects/{project_id}/scenarios/generate",
    response_model=list[Scenario],
)
def generate_scenarios(project_id: str) -> list[Scenario]:
    return _get_service().generate_scenarios(project_id)


@router.post(
    "/api/projects/{project_id}/simulations/run",
    response_model=SimulationRun,
    status_code=201,
)
def run_simulation(project_id: str, body: RunSimulationRequest) -> SimulationRun:
    return _get_service().run_simulation(project_id, body.scenarios)


@router.get(
    "/api/simulations/{simulation_id}/results",
    response_model=SimulationRun,
)
def get_results(simulation_id: str) -> SimulationRun:
    try:
        return _get_service().get_results(simulation_id)
    except KeyError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post(
    "/api/projects/{project_id}/simulations/compare",
    response_model=CompareResponse,
    status_code=201,
)
def compare_simulations(project_id: str, body: CompareRequest) -> CompareResponse:
    return _get_service().compare(project_id, body)

File: frontend/src/lib/api/simulations.ts
import { apiClient } from './client'

export type Scenario = {
  id: string
  text: string
  category: string
  vip: boolean
  legal_risk: boolean
  expected_action: string
}

export type EngineResult = {
  final_action: string
  traversed_nodes: string[]
  triggered_boundaries: string[]
  trace: string[]
}

export type ComparisonResult = {
  success: boolean
  mismatch: boolean
}

export type SimulationResult = {
  scenario: Scenario
  engine_result: EngineResult
  comparison: ComparisonResult
  trace_id: string | null
}

export type SimulationRun = {
  id: string
  project_id: string
  results: SimulationResult[]
}

export function generateScenarios(projectId: string): Promise<Scenario[]> {
  return apiClient.post<Scenario[]>(`/projects/${projectId}/scenarios/generate`, {})
}

export function runSimulation(
  projectId: string,
  scenarios: Scenario[],
): Promise<SimulationRun> {
  return apiClient.post<SimulationRun>(`/projects/${projectId}/simulations/run`, {
    scenarios,
  })
}

export function getSimulationResults(simulationId: string): Promise<SimulationRun> {
  return apiClient.get<SimulationRun>(`/simulations/${simulationId}/results`)
}
